### ISSUE TITLE
With Windows Fall Creators Update OneDriver Folder cannot be backupped any longer

### ISSUE THREAD

**Issue by hamlet42** [duplicati/duplicati#2829]
- [X ] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: v2.0.2.8-2.0.2.8_canary_2017-09-20 
- **Operating system**: Windows 10 Fall Creators Update
- **Backend**: Replace me

## Description
After upgrading to Fall Creators Update it is no longer possible to update the OneDrive Folder.
An error is logged for the respective folder
>>
Failed to process path: C:\Users\helmu\OneDrive\
Alphaleonis.Win32.Filesystem.UnrecognizedReparsePointException: E/A-Fehler

   bei Alphaleonis.Win32.Filesystem.Device.GetLinkTargetInfoCore(SafeFileHandle safeHandle)

   bei Alphaleonis.Win32.Filesystem.File.GetLinkTargetInfoCore(KernelTransaction transaction, String path, PathFormat pathFormat)

   bei Duplicati.Library.Snapshots.WindowsSnapshot.GetSymlinkTarget(String file)

   bei Duplicati.Library.Main.Operation.BackupHandler.HandleFilesystemEntry(ISnapshotService snapshot, BackendManager backend, String path, FileAttributes attributes)
>>
## Steps to reproduce
1. Include OneDrive Folder in Backup
2. Process backup. Message from above is logged
3. ...

**Comment by tygill** [comment]
I suspect this is due to OneDrive's new (well, technically restored from Windows 8.1) 'Files on Demand' feature, which lets you browse the contents of the OneDrive folder without needing to have everything downloaded locally:
https://support.office.com/en-us/article/Learn-about-OneDrive-Files-On-Demand-0e6860d3-d9f3-4971-b321-7092438fb38e

I'm not familiar with [AlphaFS](https://github.com/alphaleonis/AlphaFS) (which seems to be the library Duplicati is using to parse reparse points and other directory links), but it might require changes there to recognize this new type of reparse point.

**Comment by hamlet42** [comment]
As I was just migrated to Fall Creators Update last evening, all files are local. But nevertheless I think you are right. The error is thrown for the OneDrive folder, not any file or sub-folder in there.

**Comment by tygill** [comment]
Even if the files are cached locally, the root folder probably needs something special to support this scenario. I suspect OneDrive now does something similar to [GVFS](https://blogs.msdn.microsoft.com/devops/2017/02/03/announcing-gvfs-git-virtual-file-system/) (Git Virtual File System) to create a virtual file system and have that mounted into a particular folder.

**Comment by joyrider3774** [comment]
i'm having this problem as well after fall creators update. even on individual files.

I have the "my picutres" and "my documents" folders relocated to the one drive folder so they automatically backed up. After fall creators update i've been having such errors. The file below for example exists localy and i can open it when just pasting the filename in explorer and pressing enter

[Failed to process path: C:\Users\joyri\OneDrive\Afbeeldingen\3.PNG
Alphaleonis.Win32.Filesystem.UnrecognizedReparsePointException: I/O error occurred.
   at Alphaleonis.Win32.Filesystem.Device.GetLinkTargetInfoCore(SafeFileHandle safeHandle)
   at Alphaleonis.Win32.Filesystem.File.GetLinkTargetInfoCore(KernelTransaction transaction, String path, PathFormat pathFormat)
   at Duplicati.Library.Snapshots.WindowsSnapshot.GetSymlinkTarget(String file)
   at Duplicati.Library.Main.Operation.BackupHandler.HandleFilesystemEntry(ISnapshotService snapshot, BackendManager backend, String path, FileAttributes attributes)](url)

**Comment by macnaus** [comment]
What effect does turning on the “expert option” Symlink-policy set to Follow have?

--symlink-policy=follow

For me this seems to have rectified the OneDrive folder io errors

**Comment by hamlet42** [comment]
Does not work as a workaround Turning this option on triggers below error on multiple other folders (the error below is Access denied for "Application Data".  I am using the snapshot-policy option in  conjunction. No other expert options are set. 

The OneDrive folder is included in the backup, but to restore files did not work (no logged errors, but the files are not created in the OneDrive folder (tried to restore the files as duplicates, not to overwrite the original files) 

>>
Error reported while accessing file: C:\Users\helmu\Anwendungsdaten\
System.UnauthorizedAccessException: (5) Zugriff verweigert: [\\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy11\Users\helmu\Anwendungsdaten\*]

   bei Alphaleonis.Win32.NativeError.ThrowException(UInt32 errorCode, String readPath, String writePath)

   bei Alphaleonis.Win32.Filesystem.FindFileSystemEntryInfo.FindFirstFile(String pathLp, WIN32_FIND_DATA& win32FindData)

   bei Alphaleonis.Win32.Filesystem.FindFileSystemEntryInfo.<Enumerate>d__3`1.MoveNext()

   bei System.Linq.Buffer`1..ctor(IEnumerable`1 source)

   bei System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)

   bei Duplicati.Library.Snapshots.WindowsSnapshot.ListFiles(String folder)

   bei Duplicati.Library.Utility.Utility.<EnumerateFileSystemEntries>d__23.MoveNext()
>>

**Comment by macnaus** [comment]
My snapshot policy is set to Auto

--snapshot-policy=Auto

Apologies if that doesn't help; for me it was the solution.

**Comment by ihnorton** [comment]
See https://github.com/libuv/libuv/pull/1419 and the issues linked therein, and the correct fix at https://github.com/libuv/libuv/pull/1522. tl;dr OneDrive uses non-symlink reparse points, which breaks everyone's stuff.

**Comment by kenkendk** [comment]
I am aware of the problem. Duplicati treats all reparse-points as symbolic links, which they are not.

**Comment by erickr** [comment]
I experienced the same issue on a updated win10, by setting the 
symlink-policy to follow the duplicati cli seems to cause all files to be downloaded from onedrive.

**Comment by tygill** [comment]
Did this end up getting fixed somehow (either by Duplicati or by some change to OneDrive)? I just tried to repro it and wasn't able to.

I did note that unless I explicitly specified --symlink-policy=follow, none of the files in my OneDrive folder were actually backed up. But when I do that, it seems to find and back up all the files (and Windows reports that they're being downloaded on-demand).

(Though I haven't been able to test with VSS, since it keeps giving a FileNoFound exception without saying what file wasn't found.)

**Comment by hamlet42** [comment]
I will check if the issue persists and come back. As written above, I use VSS and in combination with --symlink-policy=follow, this caused other issues. And then the files from OneDrive could not be restored.

**Comment by hamlet42** [comment]
Checked again, original issue

**Comment by tygill** [comment]
I just tried again to repro this, using the latest canary build (1/23). I backed up a single folder (named 'Paper' with a few JPEG files in it) in my OneDrive folder, and then restored it using an admin command line. (And my understanding is that the --snapshot-policy=required parameter should force VSS to be used.)

Duplicati.CommandLine.exe backup C:\DuplicatiTest\tygill.OneDrive.Backup C:\Users\tygill --include=*OneDrive\Paper* --no-encryption --snapshot-policy=required --verbose --symlink-policy=follow
Duplicati.CommandLine.exe restore C:\DuplicatiTest\tygill.OneDrive.Backup --restore-path=C:\DuplicatiTest\tygill.OneDrive.Restore --no-encryption

All of the files which I'd backed up were indeed present in the restore folder afterwards. 

winver reports that I am indeed running the Fall Creator's Update (Windows Version 1709).

**Comment by hamlet42** [comment]
Strange indeed. I tested as part of my normal backup. After including my OneDrive folder in the backup I get this error for the OneDrive folder: 

>>
Failed to process path: C:\Users\helmu\OneDrive\
Alphaleonis.Win32.Filesystem.UnrecognizedReparsePointException: E/A-Fehler

   bei Alphaleonis.Win32.Filesystem.Device.GetLinkTargetInfoCore(SafeFileHandle safeHandle)

   bei Alphaleonis.Win32.Filesystem.File.GetLinkTargetInfoCore(KernelTransaction transaction, String path, PathFormat pathFormat)

   bei Duplicati.Library.Snapshots.WindowsSnapshot.GetSymlinkTarget(String file)

   bei Duplicati.Library.Main.Operation.BackupHandler.HandleFilesystemEntry(ISnapshotService snapshot, BackendManager backend, String path, FileAttributes attributes)
>>

When I now add the option symlink-policy=follow, the error goes away, but I now get several errors on other folders, e.g. My Music. or My Files
>>
rror reported while accessing file: C:\Users\helmu\Eigene Dateien\
System.UnauthorizedAccessException: (5) Zugriff verweigert: [\\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy12\Users\helmu\Eigene Dateien\*]

   bei Alphaleonis.Win32.NativeError.ThrowException(UInt32 errorCode, String readPath, String writePath)

   bei Alphaleonis.Win32.Filesystem.FindFileSystemEntryInfo.FindFirstFile(String pathLp, WIN32_FIND_DATA& win32FindData)

   bei Alphaleonis.Win32.Filesystem.FindFileSystemEntryInfo.<Enumerate>d__3`1.MoveNext()

   bei System.Linq.Buffer`1..ctor(IEnumerable`1 source)

   bei System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)

   bei Duplicati.Library.Snapshots.WindowsSnapshot.ListFolders(String folder)

   bei Duplicati.Library.Utility.Utility.<EnumerateFileSystemEntries>d__22.MoveNext()
>>

When I try to restore a file from my OneDrive folder, it is not created. A box shows up stating that there are zero warnings after restore completes, that's all.

One difference in testing this may be: I am backing up my whole user folder where the OneDrive folder is one sub-folder. You  tried to back up one sub-folder in OneDrive.

Other options: snapshot-policy=required. That's all what is specifically set.

Hope this helps to drill down on this. Thx for your assistance, tygill

**Comment by tygill** [comment]
Well I feel pretty silly right now. For some reason, I was expecting to see the 'UnrecognizedReparsePointException' in the command line output. Instead, it was just showing a generic 'IOException' - I didn't realize I needed to check the actual logs to see the details. Anyway, I was able to repro it.

I've started a change to address this issue, with the root cause being (as kenkendk mentioned above) that Duplicati treats all reparse points as symlinks. In this change, I check the symlink target, and only consider a reparse point to be a symlink if the target is non-null. Otherwise (e.g., in the case of the OneDrive folder), it treats it as a regular file/directory and backs it up (regardless of the --symlink-policy setting). I did some initial testing, and it seems to be working, but I'll probably try to test some more before I submit the pull request.
